### PR TITLE
Generate backward graph via functorch-aot module

### PR DESCRIPTION
**Pull request by pashu123**
Example to demonstrate the extraction of forward as well as
backward graph via Functorch's AOT module is added.

### CODE REVIEW

**Review by silvasean — CHANGES_REQUESTED** [review]
I think you need to update https://github.com/llvm/torch-mlir/blob/main/build_tools/torchscript_e2e_heavydep_tests/generate_serialized_tests.sh to install functorch.

File: examples/aot_backward.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+# Functorch is required.
+# Please refer to: https://github.com/pytorch/functorch/blob/main/README.md
```
[the remark this answers is not in the document]
Comment (pashu123) [reply examples/aot_backward.py]: Done. Thanks.

**Comment by pashu123** [comment]
@silvasean Addressed all your comments. PTAL.

File: build_tools/torchscript_e2e_heavydep_tests/aot_backward.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    min_cut_rematerialization_partition,
+)
+from torch import optim, fx
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+#### torch-mlir imports
+# To run the example, make sure the following are in your PYTHONPATH:
+# 1. /path/to/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir
```
Comment (silvasean) [inline build_tools/torchscript_e2e_heavydep_tests/aot_backward.py:15]: you can remove this comment

File: build_tools/torchscript_e2e_heavydep_tests/aot_backward.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
+from functorch.compile import (
+    memory_efficient_fusion,
```
Comment (silvasean) [inline build_tools/torchscript_e2e_heavydep_tests/aot_backward.py:2]: do we need all these imports?

File: build_tools/torchscript_e2e_heavydep_tests/aot_backward.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    f = torch.jit.freeze(f.eval())
+    torch.jit.save(f, "back.pt")
+    f = torch.jit.load("back.pt")
+    print("The backward torch-mlir graph is:")
+    generate_mlir_module(
+        f, generate_annotation(inps), "backward.mlir"
+    )
+    return f
+
+
```
Comment (silvasean) [inline build_tools/torchscript_e2e_heavydep_tests/aot_backward.py:121]: can you please check this in as an actual e2e test?

See the structure of https://github.com/llvm/torch-mlir/blob/main/build_tools/torchscript_e2e_heavydep_tests/minilm_sequence_classification.py

Notice:
```
@register_test_case(module_factory=lambda: getTracedRecursiveScriptModule())
def MiniLMSequenceClassification_basic(module, tu: TestUtils):
    module.forward(test_input)
```

Can you also please verify that you are able to run the resulting test (even if it fails) following the instructions at https://github.com/llvm/torch-mlir/tree/main/build_tools/torchscript_e2e_heavydep_tests/README.md

**Review by silvasean — CHANGES_REQUESTED** [review]

File: build_tools/torchscript_e2e_heavydep_tests/aot_backward.py (hunk 3 above)
Comment (pashu123) [reply build_tools/torchscript_e2e_heavydep_tests/aot_backward.py:2]: No, thanks for pointing this out.

File: build_tools/torchscript_e2e_heavydep_tests/aot_backward.py (hunk 4 above)
Comment (pashu123) [reply build_tools/torchscript_e2e_heavydep_tests/aot_backward.py:121]: This test just demonstrates the generation of the backward graph via AOT autograd, the intention to add it as an e2e test is not there.

Comment (silvasean) [reply build_tools/torchscript_e2e_heavydep_tests/aot_backward.py:121]: I consider it a requirement to actually add it as an e2e test. Then we can all work together towards running it which is a major project goal.

**Comment by silvasean** [comment]
Any reason we are closing this? It is definitely a project priority to run this, so I think it would be good to keep working on this.

**Comment by pashu123** [comment]
> Any reason we are closing this? It is definitely a project priority to run this, so I think it would be good to keep working on this.

Sorry for closing. I will keep working on this.

**Comment by pashu123** [comment]
@silvasean Any suggestions on where can I place the functorch_utils.py?

**Comment by pashu123** [comment]
@silvasean This may return different return types not supported by refbackend. If you could re-iterate the pointers you mentioned in the meeting to support different return types it would be a help! I will start working on this.

**Comment by pashu123** [comment]
> @silvasean @cathyzhyi The MiniLM test case is currently failing
> 
> ```
> FAIL - "MiniLMSequenceClassification_basic"                                                                                                                                                                        
>                                                                                                                                                                                                                    
> Unexpected outcome summary:                                                                              
>                                                                                                                                                                                                                    
> ****** Failed tests - 1 tests                                                                                                                                                                                      
>     FAIL - "MiniLMSequenceClassification_basic"                                                          
>         Compilation error: Traceback (most recent call last):                                                                                                                                                      
>           File "/home/prashant/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/torchscript/framework.py", line 248, in run_tests                                                  
>             compiled = config.compile(test.program_factory())                                            
>           File "/home/prashant/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/torchscript/configs/linalg_on_tensors_backend.py", line 37, in compile
>             module = convert_torchscript_module_to_torch_backend_contract_mlir(                          
>           File "/home/prashant/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/torchscript/configs/utils.py", line 75, in convert_torchscript_module_to_torch_backend_contract_mli
> r                                                                                                                                                                                                                  
>             raise Exception(f"""                                                                                                                                                                                   
>         Exception:                                                                                                                                                                                                 
>         PyTorch TorchScript module -> torch-mlir Object Graph IR import failed with:                                                                                                                               
>         Exception:                                                                                       
>         required keyword attribute 'unsafe' is undefined                                                                                                                                                           
>         Diagnostics:  
> ```
> 
> Any suggestions?

I did a fresh install and these problems were gone! Thanks to @cathyzhyi

**Comment by silvasean** [comment]
> @silvasean This may return different return types not supported by refbackend. If you could re-iterate the pointers you mentioned in the meeting to support different return types it would be a help! I will start working on this.

You can take the torch fx graph and munge it to remove the None return types.

**Comment by pashu123** [comment]
> > @silvasean This may return different return types not supported by refbackend. If you could re-iterate the pointers you mentioned in the meeting to support different return types it would be a help! I will start working on this.
> 
> You can take the torch fx graph and munge it to remove the None return types.

@silvasean Oh! Not for the None type, I have already done that in this PR. I was asking about returning multiple different tensors through ref-backend currently it supports only fixed return types.

**Comment by silvasean** [comment]
> I was asking about returning multiple different tensors through ref-backend currently it supports only fixed return types.

Oh, the idea is that the refback-munge-calling-conventions pass would embed a string listing all the "consume_return" signatures it needs, and then in Python read that string, create all the consume_return functions, and then register them here: https://github.com/llvm/torch-mlir/blob/24f9de7120bcd95de3d341f325a46d19a6a2dc4c/python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py#L105

File: build_tools/torchscript_e2e_heavydep_tests/bert_functorch.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+# Also available under a BSD-style license. See LICENSE.
+
```
Comment (silvasean) [inline build_tools/torchscript_e2e_heavydep_tests/bert_functorch.py:5]: nit: one newline only

**Pull request merged**